openhands_cli/acp_impl/event.py

```diff
@@ -0,0 +1,523 @@
+
+    Supports:
+    - file_editor: path, view_range, insert_line
+    - Other tools with 'path' or 'directory' attributes
+
+    Args:
+        action: Action to extract locations from
+
+    Returns:
+        List of ToolCallLocation objects or None
+    """
+    locations = []
+    if isinstance(action, FileEditorAction):
+        # Handle FileEditorAction specifically
+        if action.path:
+            location = ToolCallLocation(path=action.path)
+            if action.view_range and len(action.view_range) > 0:
+                location.line = action.view_range[0]
+            elif action.insert_line is not None:
+                location.line = action.insert_line
+            locations.append(location)
+    return locations if locations else None
+
+
+def _event_visualize_to_plain(event: Event) -> str:
+    """Convert Rich Text object to plain string.
+
+    Args:
+        text: Rich Text object or string
+
+    Returns:
+        Plain text string
+    """
+    text = event.visualize
+    text = text.plain
+    return str(text)
+
+
+class EventSubscriber:
+    """Subscriber for handling OpenHands events and converting them to ACP
+    notifications.
+
+    This class subscribes to events from an OpenHands conversation and converts
+    them to ACP session update notifications that are streamed back to the client.
+    """
+
+    def __init__(self, session_id: str, conn: "AgentSideConnection"):
+        """Initialize the event subscriber.
+
+        Args:
+            session_id: The ACP session ID
+            conn: The ACP connection for sending notifications
+        """
+        self.session_id = session_id
+        self.conn = conn
+
+    async def __call__(self, event: Event):
+        """Handle incoming events and convert them to ACP notifications.
+
+        Args:
+            event: Event to process (ActionEvent, ObservationEvent, etc.)
+        """
+        # Skip ConversationStateUpdateEvent (internal state management)
+        if isinstance(event, ConversationStateUpdateEvent):
+            return
+
+        # Handle different event types
+        if isinstance(event, ActionEvent):
+            await self._handle_action_event(event)
+        elif isinstance(
+            event, ObservationEvent | UserRejectObservation | AgentErrorEvent
+        ):
+            await self._handle_observation_event(event)
+        elif isinstance(event, MessageEvent):
+            await self._handle_message_event(event)
+        elif isinstance(event, SystemPromptEvent):
+            await self._handle_system_prompt_event(event)
+        elif isinstance(event, PauseEvent):
+            await self._handle_pause_event(event)
+        elif isinstance(event, Condensation):
+            await self._handle_condensation_event(event)
+        elif isinstance(event, CondensationRequest):
+            await self._handle_condensation_request_event(event)
+
+    async def _handle_action_event(self, event: ActionEvent):
```

Comment:
how does this interact with parallel tool calls? does the client invoke this method on one tool call at a time?